pkg/controller/podgroup.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2023 The Kubeflow Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"sort"
+
+	"github.com/google/go-cmp/cmp"
+	common "github.com/kubeflow/common/pkg/apis/common/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+	volcanov1beta1 "volcano.sh/apis/pkg/apis/scheduling/v1beta1"
+
+	kubeflow "github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v2beta1"
+)
+
+type replicaPriority struct {
+	priority int32
+
+	common.ReplicaSpec
+}
+
+// newPodGroup creates a new PodGroup for an MPIJob
+// resource. It also sets the appropriate OwnerReferences on the resource so
+// handleObject can discover the MPIJob resource that 'owns' it.
+func (c *MPIJobController) newPodGroup(mpiJob *kubeflow.MPIJob) *volcanov1beta1.PodGroup {
+	minMember := workerReplicas(mpiJob) + 1
+	queueName := mpiJob.Annotations[volcanov1beta1.QueueNameAnnotationKey]
+	var minResources *corev1.ResourceList
+	if schedulingPolicy := mpiJob.Spec.RunPolicy.SchedulingPolicy; schedulingPolicy != nil {
+		if schedulingPolicy.MinAvailable != nil {
+			minMember = *schedulingPolicy.MinAvailable
```

Comment:
> what if volcano schedules a number of pods that is less than workerReplicas(mpiJob)?
The mpirun command might get stuck waiting for all the workers (unless it's elastic horovod that has native support for changing number of workers).

In that case, volcano schedules all Pods. If the cluster does not have enough resources for all Pods, some pods will be marked as `Pending`.

> We need a way to restrict the usage of this field or have clear documentation that this is not supported for all kinds of MPI applications.

I agree. I think we have 2 options. Since we don't have fields if the MPIJob uses elastic horovod. 

1. As you say, have clear documentation.
2. Add a new member, 'Elastic bool `json:elastic,omitempty`' into `Runpolicy` to represent if MPIJob uses elastic mode. And then, we add validations for `minMember` and `minResources`.

I like the second option. What do you think? @alculquicondor

Reply:
Are there similar fields in the training operator?

Reply:
No, most customjob-controllers have no fields on whether CustomJob uses elastic mode. This means users can set any value to `schedulingPolicy.MinMember`. The pytorchjob-controller has `ElasticPolicy`, although that field doesn't affect schedulingPolicy.

https://github.com/kubeflow/training-operator/blob/6ce98387a1a68b79365495a6344f7f19b48557d3/pkg/apis/kubeflow.org/v1/pytorch_types.go#L79

Reply:
in that case, I think we should proceed with option 1 for now.

Reply:
> in that case, I think we should proceed with option 1 for now.

I agree. Maybe, we can add documentation to https://www.kubeflow.org/docs/components/training/job-scheduling/.

> That said, there is currently no documentation about SchedulingPolicy overall 

You're right...

Reply:
> So we need to clarify it in the documentation anyways.

I agree.

Probably, we can discuss if we add an `elastic` field in other places (issue or pr).
In this PR, we can just pass `SchedulingPolicy.MinAvailable` to `PodGroup.Spec.MinMember` and then add docs about that.